### PR TITLE
core: prng: Don't lock a mutex from rpc_cmd_nolock()

### PR DESCRIPTION
**Pull request by jenswi-linaro**
Obsoletes https://github.com/OP-TEE/optee_os/pull/1385

### CODE REVIEW

**Comment by etienne-lms** [comment]
Nice catch indeed.

Regarding the debug part, could you replace the owner IDs -1, -2 -3 with some macro. These mutex stuff are quite hard to follow. Having `OWNER_ID_NONE` instead of -1, `OWNER_ID_MUTEX_WAKER` instead of -2, `OWNER_ID_MUTEX_???` instead of -3 would help (i did not catch the condvar wait/wake sequence). I think.

I'll stress test my b2260, enabling many threads to increase secure waiters...

I wonder if this recursive call could have been caught by jerome's profiling support ?

**Comment by jforissier** [comment]
> I wonder if this recursive call could have been caught by jerome's profiling support ?

No, because profiling applies to user mode code only (TAs).

File: core/arch/arm/kernel/thread.c [hunk 1]
```diff
@@ -1199,7 +1199,8 @@ static uint32_t rpc_cmd_nolock(uint32_t cmd, size_t num_params,
 
 	assert(arg && carg && num_params <= THREAD_RPC_MAX_NUM_PARAMS);
 
-	plat_prng_add_jitter_entropy_norpc();
+	if (cmd != OPTEE_MSG_RPC_CMD_WAIT_QUEUE)
```
Comment (jforissier) [inline core/arch/arm/kernel/thread.c]: Ha! That's a tricky one. I think this deserves a comment like "Break recursion in case plat_prng_add_jitter_entropy_norpc() invokes a mutex" or something like that.

Also I would recommend commenting the prototype for plat_prng_add_jitter_entropy_norpc():
```diff
diff --git a/core/include/tee/tee_cryp_utl.h b/core/include/tee/tee_cryp_utl.h
index 28b501d..76e852a 100644
--- a/core/include/tee/tee_cryp_utl.h
+++ b/core/include/tee/tee_cryp_utl.h
@@ -50,6 +50,10 @@ TEE_Result tee_aes_cbc_cts_update(void *cbc_ctx, void *ecb_ctx,
 
 TEE_Result tee_prng_add_entropy(const uint8_t *in, size_t len);
 void plat_prng_add_jitter_entropy(void);
+/*
+ * The _norpc version must not invoke Normal World, or infinite recursion
+ * may occur. As an exception however, using mutexes is allowed.
+ */
 void plat_prng_add_jitter_entropy_norpc(void);
 
 #endif
```

**Comment by etienne-lms** [comment]
> > I wonder if this recursive call could have been caught by jerome's profiling support ?
> >
> No, because profiling applies to user mode code only (TAs).

oups! right.

**Comment by jenswi-linaro** [comment]
Update

**Comment by etienne-lms** [comment]
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`

**Comment by jforissier** [comment]
For "core: prng: Don't lock a mutex from rpc_cmd_nolock()":
`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

**Comment by jenswi-linaro** [comment]
Rebased, squashed, tags applied

**Comment by jbech-linaro** [comment]
Thanks Jens / all for the quick turn around time here!
